shared/src/ledger/storage/merkle_tree.rs

```diff
@@ -0,0 +1,550 @@
+    NonExistenceProof, ProofSpec,
+};
+use prost::Message;
+use sha2::{Digest, Sha256};
+use sparse_merkle_tree::default_store::DefaultStore;
+use sparse_merkle_tree::error::Error as SmtError;
+use sparse_merkle_tree::traits::Hasher;
+use sparse_merkle_tree::{SparseMerkleTree, H256};
+#[cfg(not(feature = "ABCI"))]
+use tendermint::merkle::proof::{Proof, ProofOp};
+#[cfg(feature = "ABCI")]
+use tendermint_stable::merkle::proof::{Proof, ProofOp};
+use thiserror::Error;
+
+use crate::bytes::ByteBuf;
+use crate::types::address::{Address, InternalAddress};
+use crate::types::storage::{DbKeySeg, Error as StorageError, Key};
+
+#[allow(missing_docs)]
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Invalid key: {0}")]
+    InvalidKey(StorageError),
+    #[error("Empty Key: {0}")]
+    EmptyKey(String),
+    #[error("SMT error: {0}")]
+    Smt(SmtError),
+    #[error("Proof spec error")]
+    InvalidProofSpec,
+}
+
+/// Result for functions that may fail
+type Result<T> = std::result::Result<T, Error>;
+
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    Hash,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    BorshSerialize,
+    BorshDeserialize,
+)]
+pub enum StoreType {
+    Base,
+    Account,
+    PoS,
+    Ibc,
+}
+
+impl StoreType {
+    fn sub_tree_iter() -> std::slice::Iter<'static, Self> {
+        static SUB_TREE_TYPES: [StoreType; 3] =
+            [StoreType::Account, StoreType::PoS, StoreType::Ibc];
+        SUB_TREE_TYPES.iter()
+    }
+
+    fn sub_key(key: &Key) -> Result<(Self, Key)> {
+        if key.is_empty() {
+            return Err(Error::EmptyKey("the key is empty".to_owned()));
+        }
+        match key.segments.get(0) {
+            Some(DbKeySeg::AddressSeg(Address::Internal(internal))) => {
+                match internal {
+                    InternalAddress::PoS | InternalAddress::PosSlashPool => {
+                        Ok((StoreType::PoS, key.sub_key()?))
+                    }
+                    InternalAddress::Ibc => {
+                        Ok((StoreType::Ibc, key.sub_key()?))
+                    }
+                    // use the same key for Parameters
+                    _ => Ok((StoreType::Account, key.clone())),
+                }
+            }
+            // use the same key for Account
+            _ => Ok((StoreType::Account, key.clone())),
+        }
+    }
+}
+
+impl fmt::Display for StoreType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            StoreType::Base => write!(f, "base"),
+            StoreType::Account => write!(f, "account"),
+            StoreType::PoS => write!(f, "pos"),
+            StoreType::Ibc => write!(f, "ibc"),
+        }
+    }
+}
+
+/// Merkle tree storage
+pub struct MerkleTree<H: StorageHasher + Default> {
+    base: SparseMerkleTree<H, H256, DefaultStore<H256>>,
+    subtrees: HashMap<StoreType, SparseMerkleTree<H, H256, DefaultStore<H256>>>,
```

Comment:
I'm not sure if we really need this HashMap as we'll always have static number of the subtrees. I'd be in favor of adding a field for each subtree type (like for the `base`) instead